File: apps/api/internal/email/ics.go
package email

import (
	"fmt"
	"strings"
	"time"

	"flowbook/api/internal/db"
	"github.com/google/uuid"
)

// BuildICS generates RFC5545 iCalendar content for a booking.
// Times are stored UTC in DB and rendered as UTC Zulu per spec (DTSTART:UTC).
// ORGANIZER from resendFrom, ATTENDEE is customer.
func BuildICS(booking db.Booking, service db.Service, staff db.Staff, org db.Organization) string {
	// UID must be globally unique per booking
	uid := fmt.Sprintf("%s@flowbook.example.com", booking.ID.String())
	// DTSTAMP is now UTC
	now := time.Now().UTC().Format("20060102T150405Z")
	dtStart := booking.StartAt.UTC().Format("20060102T150405Z")
	dtEnd := booking.EndAt.UTC().Format("20060102T150405Z")
	// Escape text per RFC5545
	escape := func(s string) string {
		r := strings.ReplaceAll(s, "\\", "\\\\")
		r = strings.ReplaceAll(r, ";", "\\;")
		r = strings.ReplaceAll(r, ",", "\\,")
		r = strings.ReplaceAll(r, "\n", "\\n")
		return r
	}
	summary := escape(fmt.Sprintf("%s with %s", service.Name, staff.Name))
	location := escape(org.Name)
	if org.Name == "" {
		location = escape("FlowBarber Studio")
	}
	descriptionParts := []string{
		fmt.Sprintf("Booking %s", booking.ID.String()),
		fmt.Sprintf("Service: %s (%d min)", service.Name, service.DurationMinutes),
		fmt.Sprintf("Staff: %s", staff.Name),
		fmt.Sprintf("Customer: %s <%s>", booking.CustomerName, booking.CustomerEmail),
	}
	if booking.Notes.Valid && strings.TrimSpace(booking.Notes.String) != "" {
		descriptionParts = append(descriptionParts, fmt.Sprintf("Notes: %s", booking.Notes.String))
	}
	descriptionParts = append(descriptionParts, fmt.Sprintf("View: /track/%s", booking.ID.String()))
	description := escape(strings.Join(descriptionParts, "\\n"))

	// Status mapping: CONFIRMED vs others
	status := "CONFIRMED"
	if booking.Status == "CANCELLED" {
		status = "CANCELLED"
	}

	// ORGANIZER from org or default
	organizer := "noreply@flowbook.example.com"
	if strings.Contains(org.Name, "@") {
		organizer = org.Name
	}

	return strings.Join([]string{
		"BEGIN:VCALENDAR",
		"VERSION:2.0",
		"PRODID:-//FlowBook//Booking//EN",
		"CALSCALE:GREGORIAN",
		"METHOD:REQUEST",
		"BEGIN:VEVENT",
		fmt.Sprintf("UID:%s", uid),
		fmt.Sprintf("DTSTAMP:%s", now),
		fmt.Sprintf("DTSTART:%s", dtStart),
		fmt.Sprintf("DTEND:%s", dtEnd),
		fmt.Sprintf("SUMMARY:%s", summary),
		fmt.Sprintf("DESCRIPTION:%s", description),
		fmt.Sprintf("LOCATION:%s", location),
		fmt.Sprintf("STATUS:%s", status),
		fmt.Sprintf("ORGANIZER;CN=%s:mailto:%s", escape(org.Name), organizer),
		fmt.Sprintf("ATTENDEE;CN=%s;RSVP=TRUE:mailto:%s", escape(booking.CustomerName), booking.CustomerEmail),
		fmt.Sprintf("SEQUENCE:%d", 0),
		"END:VEVENT",
		"END:VCALENDAR",
		"",
	}, "\r\n")
}

// BuildICSUID returns deterministic UID for a booking (for testing).
func BuildICSUID(bookingID uuid.UUID) string {
	return fmt.Sprintf("%s@flowbook.example.com", bookingID.String())
}

File: apps/web/lib/ics.ts
import type { Booking, Service } from "./api";

function toICSDate(iso: string): string {
  const d = new Date(iso);
  // UTC format YYYYMMDDTHHMMSSZ
  const pad = (n: number) => String(n).padStart(2, "0");
  return (
    d.getUTCFullYear().toString() +
    pad(d.getUTCMonth() + 1) +
    pad(d.getUTCDate()) +
    "T" +
    pad(d.getUTCHours()) +
    pad(d.getUTCMinutes()) +
    pad(d.getUTCSeconds()) +
    "Z"
  );
}

export function buildICS(booking: Booking, service?: Service): string {
  const now = toICSDate(new Date().toISOString());
  const dtStart = toICSDate(booking.startAt);
  const dtEnd = toICSDate(booking.endAt);
  const summary = service ? `${service.name} — FlowBarber Studio` : "Booking FlowBarber Studio";
  const description = `Booking ${booking.id}\\nStaff: ${booking.staffId}\\nCustomer: ${booking.customerName} (${booking.customerEmail})\\n${booking.notes ?? ""}`.replace(/\n/g, "\\n");
  const location = "FlowBarber Studio, Jakarta";
  const uid = `${booking.id}@flowbook.example.com`;

  return [
    "BEGIN:VCALENDAR",
    "VERSION:2.0",
    "PRODID:-//FlowBook//Booking//ID",
    "CALSCALE:GREGORIAN",
    "METHOD:PUBLISH",
    "BEGIN:VEVENT",
    `UID:${uid}`,
    `DTSTAMP:${now}`,
    `DTSTART:${dtStart}`,
    `DTEND:${dtEnd}`,
    `SUMMARY:${summary}`,
    `DESCRIPTION:${description}`,
    `LOCATION:${location}`,
    "STATUS:CONFIRMED",
    "END:VEVENT",
    "END:VCALENDAR",
  ].join("\r\n");
}

export function downloadICS(booking: Booking, service?: Service) {
  const ics = buildICS(booking, service);
  const blob = new Blob([ics], { type: "text/calendar;charset=utf-8" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `flowbook-${booking.id}.ics`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}
